Scatter matrix

Casyfill opened the pull request.

pratapvardhan commented:
I guess, we'll not be showing kde plots on the diagonal. Should we add a note?

Casyfill commented:
true, but we're not substituting pandas scatter_matrix per seu (see #24).

But it definitely makes sense to add a docstring anyway

Casyfill commented:
Added some documentation, including the caveat on kde

Casyfill commented:
Happy to drop support for custom tooltips, if anyone has a strong feeling - in defense of the existing approach, we're "explicitly" not overriding pandas scatter_matrix, so there is no chance people would assume those are 100% interchangable

jakevdp commented on altair_pandas/_misc.py:
```diff
@@ -0,0 +1,64 @@
+        else:
+            color = alt.value(color)
+    elif hasattr(color, "__len__") and len(color) == len(df):
+        dfc["color"] = color
```
Worth checking if "color" is an existing column name to avoid overwriting it.

jakevdp commented on altair_pandas/test_plotting.py:
```diff
@@ -227,3 +228,51 @@ def test_dataframe_area(dataframe, stacked, with_plotting_backend):
+        assert set(el["field"] for el in spec["spec"]["encoding"]["tooltip"]) == {
+            "x",
+            "y",
+        }
```
Probably worth changing the tooltip inputs so this case is different than the default.

Casyfill replied:
sure!

pratapvardhan commented on altair_pandas/_misc.py:
```diff
@@ -0,0 +1,73 @@
+    if color:
+        chart = chart.encode(color=color)
+
+    return chart.repeat(row=cols, column=cols[::-1]).interactive()
```
`chart.repeat(row=cols, column=cols)` can be same I suppose? [matplotlib](https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html#scatter-matrix-plot) has them on other diagonal.

Casyfill replied:
There is an Altair bug (?) that prevents from doing that :-)
Note that the same is done in Altair example:

https://altair-viz.github.io/gallery/scatter_matrix.html

pratapvardhan replied:
Ah, didn't know about that. `chart.repeat(row=cols, column=cols)` seems to work for me, on my versions..

jakevdp replied:
I don't think there's any bug - it just decides the orientation of the diagonals

Casyfill replied:
Hm. here is my Issue (am I missing something?):

```
df = pd.DataFrame(np.random.randn(100,3).cumsum(0), columns=['x','y','z'])
cols = df.columns.tolist()


alt.Chart(df).mark_circle().encode(
    alt.X(alt.repeat("column"), type='quantitative'),
    alt.Y(alt.repeat("row"), type='quantitative'),
    color='Origin:N'
).properties(
    width=150,
    height=150
).repeat(
    row=cols,
    column=cols
).interactive()
```

This results in this error:
```
Javascript Error: Expression parse error: {"x": selector003_x, "x": selector003_x_1, "z": selector003_z}
```
Altair 2.3.0

jakevdp replied:
What frontend renderer are you using?

Casyfill replied:
renderer jupyterlab
jupyterlab                1.0.2

Casyfill commented:
There is an Altair bug (?) that prevents from doing that :-)

On Sat, Aug 31, 2019 at 11:38 AM Pratap Vardhan <notifications@github.com>
wrote:

> *@pratapvardhan* commented on this pull request.
> ------------------------------
>
> In altair_pandas/_misc.py
> <https://github.com/altair-viz/altair_pandas/pull/31#discussion_r319727993>
> :
>
> > +    chart = (
> +        alt.Chart(dfc)
> +        .mark_circle()
> +        .encode(
> +            x=alt.X(alt.repeat("column"), type="quantitative"),
> +            y=alt.X(alt.repeat("row"), type="quantitative"),
> +            opacity=alt.value(alpha),
> +            tooltip=tooltip,
> +        )
> +        .properties(width=150, height=150)
> +    )
> +
> +    if color:
> +        chart = chart.encode(color=color)
> +
> +    return chart.repeat(row=cols, column=cols[::-1]).interactive()
>
> chart.repeat(row=cols, column=cols) can be same I suppose? matplotlib
> <https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html#scatter-matrix-plot>
> has them on other diagonal.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/altair-viz/altair_pandas/pull/31?email_source=notifications&email_token=AAUNHJLQN7JAVLHMK3OLMWLQHKGAJA5CNFSM4ISOST22YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDJ3PEQ#pullrequestreview-282310546>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAUNHJNKGME7WACRPWA3MIDQHKGAJANCNFSM4ISOST2Q>
> .
>
-- 
Philipp Kats,
data journalism, analysis and visualization

philipp-kats.squarespace.com <http://cargocollective.com/Cityfish/>


email: casyfill@gmail.com
cellphone: +7(916)438 68 04

jakevdp commented:
(The version of Altair doesn't matter so much as the version of Vega-Lite, which is determined by the renderer)

pratapvardhan commented on altair_pandas/_misc.py:
```diff
@@ -0,0 +1,73 @@
+        DataFame to be used for scatterplot. Only numeric columns will be included.
+    color : string or iterable [optional]
+        Can be a column name, specific color value (hex, webcolors), or an array
+        of values to be used.
```
I was wondering, if we need to support `array of values`. Use can always pass precreated dataframe like `scatter_matrix(df.assign(colorcol=array_values), ... )` -- having to handle column name or color value would be much simpler/cleaner?

Casyfill commented:
But then it will be shown in the scatter matrix?

On Sun, Sep 1, 2019 at 10:09 AM Pratap Vardhan <notifications@github.com>
wrote:

> *@pratapvardhan* commented on this pull request.
> ------------------------------
>
> In altair_pandas/_misc.py
> <https://github.com/altair-viz/altair_pandas/pull/31#discussion_r319761669>
> :
>
> > +    alpha: float = 1.0,
> +    tooltip: Union[List[str], toltipList, None] = None,
> +):
> +    """ plots a scatter matrix
> +
> +    At the moment does not support neither histogram nor kde;
> +    Uses f-f scatterplots instead. Interactive and with a cusotmizable
> +    tooltip
> +
> +    Parameters
> +    ----------
> +    df : DataFame
> +        DataFame to be used for scatterplot. Only numeric columns will be included.
> +    color : string or iterable [optional]
> +        Can be a column name, specific color value (hex, webcolors), or an array
> +        of values to be used.
>
> I was wondering, if we need to support array of values. Use can always
> pass precreated dataframe like scatter_matrix(df.assign(colorcol=array_values),
> ... ) -- having to handle column name or color value would be much
> simpler/cleaner?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/altair-viz/altair_pandas/pull/31?email_source=notifications&email_token=AAUNHJOW724ECSCGY6QBPC3QHPEKHA5CNFSM4ISOST22YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCDKEHFI#pullrequestreview-282346389>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAUNHJK3HSZF4H7PFABHCFLQHPEKHANCNFSM4ISOST2Q>
> .
>
-- 
Philipp Kats,
data journalism, analysis and visualization

philipp-kats.squarespace.com <http://cargocollective.com/Cityfish/>


email: casyfill@gmail.com
cellphone: +7(916)438 68 04

pratapvardhan commented:
Only if `color` column is numeric? Since `cols = dfc._get_numeric_data().columns.astype(str).tolist()`.

Casyfill commented:
Yes, but still

On Sun, Sep 1, 2019 at 10:17 AM Pratap Vardhan <notifications@github.com>
wrote:

> Only if color column is numeric? Since cols =
> dfc._get_numeric_data().columns.astype(str).tolist().
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/altair-viz/altair_pandas/pull/31?email_source=notifications&email_token=AAUNHJLCZCIYO4FFQKQ6ZTTQHPFF3A5CNFSM4ISOST22YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5UDIWI#issuecomment-526922841>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAUNHJNLHOUHINQYE6VAHRDQHPFF3ANCNFSM4ISOST2Q>
> .
>
-- 
Philipp Kats,
data journalism, analysis and visualization

philipp-kats.squarespace.com <http://cargocollective.com/Cityfish/>


email: casyfill@gmail.com
cellphone: +7(916)438 68 04

Casyfill commented:
> Only if `color` column is numeric? Since `cols = dfc._get_numeric_data().columns.astype(str).tolist()`.

Yes, but still

Casyfill commented:
@pratapvardhan you're probably right and this case of numeric-out-of-df array is a theoretical one; Will drop that option

Casyfill commented:
This version supports colormap AND does not support color as an array

Casyfill commented:
any other feedback?

pratapvardhan commented:
@Casyfill -- I was thinking, if we should handle non-string column names for color, and would it be simpler to move the color part below like

```python

dfc = df.rename(columns=str).copy()
cols = dfc._get_numeric_data().columns.tolist()

chart = (
    alt.Chart(dfc)
    .mark_circle()
    .encode(
        x=alt.X(alt.repeat("column"), type="quantitative"),
        y=alt.X(alt.repeat("row"), type="quantitative"),
        opacity=alt.value(alpha),
        tooltip=tooltip or dfc.columns.tolist(),
    )
    .properties(width=150, height=150)
)

if color:
    color = str(color)
    if  color in dfc:
        color = alt.Color(color)
        if "colormap" in kwargs:
            color.scale = alt.Scale(scheme=kwargs.get("colormap"))
    else:
        color = alt.value(color)
    chart = chart.encode(color=color)

return chart.repeat(row=cols, column=cols).interactive()
```

What do you think?

Casyfill commented:
Thanks, @pratapvardhan That makes a lot of sense! 
Didn't know the `if  color in dfc` syntaxis.

Changed code per your suggestion, also added support for multiindex

Casyfill commented:
...

jakevdp commented:
Looking good - let me know when this is ready for a final review

Casyfill commented:
Thanks! I think it has everything I meant to add at this point

jakevdp commented on altair_pandas/_misc.py:
```diff
@@ -0,0 +1,73 @@
+from typing import Union, List
+import pandas as pd
+
+toltipList = List[alt.Tooltip]
```
tooltipList

jakevdp commented on altair_pandas/_misc.py:
```diff
@@ -0,0 +1,73 @@
+            x=alt.X(alt.repeat("column"), type="quantitative"),
+            y=alt.X(alt.repeat("row"), type="quantitative"),
+            opacity=alt.value(alpha),
+            tooltip=tooltip or dfc.columns.tolist(),
```
Seems like the contents of ``tooltip`` should be converted to strings in the same way the columns are within ``_preprocess_data``

Casyfill replied:
added a sanitizer for a tooltip, although there might be a better way to do that

jakevdp commented:
Looks good, thanks!
